cpp/src/ice2slice/Gen.cpp

```diff
@@ -0,0 +1,721 @@
+
+        ContainerPtr c = p->container();
+        p = ContainedPtr::dynamicCast(c); // This cast fails for Unit.
+        if (!p)
+        {
+            break;
+        }
+    }
+
+    assert(m);
+
+    static const string prefix = "cs:namespace:";
+
+    string q;
+    if (m->findMetaData(prefix, q))
+    {
+        hasCSharpNamespaceAttribute = true;
+        return q.substr(prefix.size()) + "." + csharpNamespace;
+    }
+    else
+    {
+        hasCSharpNamespaceAttribute = false;
+        return csharpNamespace;
+    }
+}
+
+static string getOutputName(const string& fileBase, const string& scoped, bool includeNamespace)
+{
+    ostringstream os;
+    os << fileBase;
+    if (includeNamespace)
+    {
+        assert(scoped[0] == ':');
+        string::size_type next = 0;
+        string::size_type pos;
+        while ((pos = scoped.find("::", next)) != string::npos)
+        {
+            pos += 2;
+            if (pos != scoped.size())
+            {
+                string::size_type endpos = scoped.find("::", pos);
+                if (endpos != string::npos && endpos > pos)
+                {
+                    os << "_" << scoped.substr(pos, endpos - pos);
+                }
+            }
+            next = pos;
+        }
+
+        if (next != scoped.size())
+        {
+            os << "_" << scoped.substr(next);
+        }
+    }
+    os << ".slice";
+    return os.str();
+}
+
+string getUnqualified(const ContainedPtr& contained, const string& moduleName)
+{
+    const string scopedName = contained->scoped();
+    if (scopedName.find("::") != string::npos &&
+        scopedName.find(moduleName) == 0 &&
+        scopedName.find("::", moduleName.size()) == string::npos)
+    {
+        return scopedName.substr(moduleName.size());
+    }
+    else
+    {
+        return scopedName;
+    }
+}
+
+string typeToString(const TypePtr& type, const string& scope, bool optional)
+{
+    ostringstream os;
+
+    static const char* builtinTable[] =
+    {
+        "uint8",                // KindByte
+        "bool",                 // KindBool
+        "int16",                // KindShort
+        "int32",                // KindInt
+        "int64",                // KindLong
+        "float32",              // KindFloat
+        "float64",              // KindDouble
+        "string",               // KindString
+        "???",                  // KindObject
+        "::Ice::ObjectProxy",   // KindObjectProxy
+        "???",                  // KindLocalObject
+        "???"                   // KindValue
+    };
+
+    BuiltinPtr builtin = BuiltinPtr::dynamicCast(type);
+    if (builtin)
+    {
+        os << builtinTable[builtin->kind()];
+    }
+
+    ContainedPtr contained = ContainedPtr::dynamicCast(type);
+    if (contained)
+    {
+        os << getUnqualified(contained, scope);
+    }
+
+    if (optional)
+    {
+        os << "?";
+    }
+
+    return os.str();
+}
+
+string typeToCsString(const TypePtr& type, bool optional)
```

Comment:
Not clear why we need this in ice2slice.

Reply:
To map cs::generic metadata to cs::type